core/rpc/client/gateway/jsonrpc/client.go

```diff
@@ -0,0 +1,188 @@
+	if rtp := reflect.TypeOf(res); rtp.Kind() != reflect.Ptr {
+		return errors.New("result must be a pointer")
+	}
+
+	// Marshal the params.
+	params, err := json.Marshal(cmd)
+	if err != nil {
+		return err
+	}
+
+	// Marshal the request.
+	id := cl.nextReqID()
+	req := jsonrpc.NewRequest(id, method, params)
+
+	request, err := json.Marshal(req)
+	if err != nil {
+		return err
+	}
+
+	// Build and perform the http request.
+	requestReader := bytes.NewReader(request)
+	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost,
+		cl.endpoint, requestReader)
+	if err != nil {
+		return fmt.Errorf("failed to construct new http request: %w", err)
+	}
+
+	httpReq.Header.Set("Content-Type", "application/json")
+	// httpReq.SetBasicAuth(c.User, c.Pass)
+
+	httpResponse, err := cl.conn.Do(httpReq)
+	if err != nil {
+		return fmt.Errorf("http post failed: %w", err)
+	}
+	defer httpResponse.Body.Close()
+
+	// For the most part we ignore the http status code in favor of structured
+	// errors in the response, but in case we cannot decode any response body,
+	// get an error based on the http status code.
+	var httpErr error
+	switch status := httpResponse.StatusCode; status {
+	case http.StatusOK: // expected with nil resp.Error
+	case http.StatusUnauthorized:
+		httpErr = rpcClient.ErrUnauthorized
+	case http.StatusNotFound:
+		httpErr = rpcClient.ErrNotFound
+	case http.StatusInternalServerError:
+		httpErr = errors.New("server error")
+	default:
+		if status >= 400 {
+			httpErr = errors.New(http.StatusText(status))
+		}
+	}
+
+	resp := &jsonrpc.Response{}
+	err = json.NewDecoder(httpResponse.Body).Decode(resp)
+	if err != nil {
+		if httpErr != nil {
+			return httpErr
+		}
+		return fmt.Errorf("failed to decode response: %w", errors.Join(err, httpErr))
+	}
+
+	if resp.Error != nil {
+		return clientError(resp.Error)
+	} // any not OK http status code should have
+
+	if resp.JSONRPC != "2.0" { // indicates response body was not a jsonrpc.Response but didn't fail Decode
+		if httpErr != nil {
+			return httpErr
+		}
+		return fmt.Errorf("invalid JSON-RPC response")
+	}
+
+	// if resp.ID != id {
+	// 	fmt.Printf("got id %v, expected %v\n", resp.ID, id)
+	// } // who cares, this is http post
+
+	if err = json.Unmarshal(resp.Result, res); err != nil {
+		return fmt.Errorf("failed to decode result as response: %w", errors.Join(err, httpErr))
+	}
+
+	return nil
+}
+
+// clientError joins a jsonrpc.Error with a client.RPCError and any appropriate
+// named error kind like ErrNotFound, ErrUnauthorized, etc. based on the code.
+func clientError(jsonRPCErr *jsonrpc.Error) error {
+	// TODO: make clientError from core/rpc/client/user/jsonrpc/client.clientError reusable
+	// Then we will first check if we recognize the error(KGW) code, if not,
+	// we use client.clientError
+
+	rpcErr := &rpcClient.RPCError{
+		Msg:  jsonRPCErr.Message,
+		Code: int32(jsonRPCErr.Code),
+	}
+	err := errors.Join(jsonRPCErr, rpcErr)
+
+	switch jsonRPCErr.Code {
```

Comment:
It seems we can delete this block(in every client)? IMO the `jsonRPCErr.Code` is very specific, and more useful for app? @jchappelow

Reply:
I think that the highest level clients want more abstraction i.e. shouldn't have to know about the jsonrpc package and codes at all.  Also `errors.Is(err, <some error kind in core/client package>)` is better for a user of `core/client.Client` I think.

Reply:
Codes could move up to a package that's not jsonrpc specific?  Or core/client could give a error code getter to extract one from this opaque `error`?

Reply:
Oh. this is in the jsonrpc package.  Hmm.  I guess core/client should be checking codes an translating to Go error types...

Reply:
You mean in core/client and core/gatewayclient ? yeah maybe, I just see those errors are not used much really, most of the time we just pop it up

Reply:
Haven't gotten a solid solution yet, I'll leave it for now